### PR TITLE
[SDK] Async flow as a func

### PR DESCRIPTION
**Pull request by crazygao**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/setup.py [hunk 1]
```diff
@@ -53,6 +53,7 @@
     "opentelemetry-exporter-otlp-proto-http>=1.22.0,<2.0.0",  # trace support
     "flask-restx>=1.2.0,<2.0.0",  # PFS Swagger
     "flask-cors>=4.0.0,<5.0.0",  # handle PFS CORS
+    "nest_asyncio",
```
Comment (wangchao1230) [inline src/promptflow/setup.py:56]: why do we need a new dep?

File: src/promptflow/promptflow/_sdk/entities/_flow.py [hunk 2]
```diff
@@ -190,11 +196,24 @@ def load(
                 data = load_yaml_string(flow_content)
                 content_hash = hash(flow_content)
             is_eager_flow = cls._is_eager_flow(data)
+            is_async_flow = cls._is_async_flow(kwargs)
             if is_eager_flow:
-                return EagerFlow._load(path=flow_path, data=data, raise_error=raise_error, **kwargs)
+                obj = EagerFlow._load(path=flow_path, data=data, raise_error=raise_error, **kwargs)
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/entities/_flow.py:201]: Shall we add a new AsyncEagerFlow class like 
```
class AsyncFlow:
    async def __call__():
       ....
```

File: src/promptflow/tests/sdk_cli_test/e2etests/test_flow_as_func.py [hunk 3]
```diff
@@ -33,6 +33,23 @@ def test_flow_as_a_func(self):
         assert result["output"] is None
         assert "line_number" not in result
 
+    @pytest.mark.asyncio
+    async def test_flow_as_a_func_asynckw(self):
+        f = load_flow(f"{FLOWS_DIR}/print_env_var", async_call=True)
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_test/e2etests/test_flow_as_func.py:38]: may add test for both dag and eager flow

File: src/promptflow/tests/sdk_cli_test/e2etests/test_flow_as_func.py [hunk 4]
```diff
@@ -33,6 +33,23 @@ def test_flow_as_a_func(self):
...
+        assert result["output"] is not None
+
+    @pytest.mark.asyncio
+    async def test_async_flow_as_a_func_asynckw(self):
+        f = load_flow(f"{FLOWS_DIR}/print_env_var_async", async_call=True)
+        result = await f(key="PATH")
+        assert result["output"] is not None
+
+    def test_async_flow_as_a_func_wo_asynckw(self):
+        f = load_flow(f"{FLOWS_DIR}/print_env_var_async")
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_test/e2etests/test_flow_as_func.py:49]: do we define async flow as dag flow that use async tools?

Then could we run flow with sync tools in async manner？

File: src/promptflow/promptflow/_sdk/entities/_flow.py [hunk 5]
```diff
@@ -190,11 +196,24 @@ def load(
...
             else:
                 # TODO: schema validation and warning on unknown fields
-                return ProtectedFlow._load(path=flow_path, dag=data, content_hash=content_hash, **kwargs)
+                obj = ProtectedFlow._load(path=flow_path, dag=data, content_hash=content_hash, **kwargs)
```
Comment (D-W-) [inline src/promptflow/promptflow/_sdk/entities/_flow.py:204]: The `load_flow(async_call=False)` will return `Flow`
But `load_flow(async_call=True)` will return a callable.


flow = load_flow(async_call=True)
flow.xxx = xx

File: src/promptflow/tests/sdk_cli_test/e2etests/test_flow_as_func.py (hunk 4 above)
Comment (crazygao) [reply src/promptflow/tests/sdk_cli_test/e2etests/test_flow_as_func.py:49]: Yes, the first test is sync tool run in async manner.

File: src/promptflow/promptflow/_sdk/entities/_flow.py (hunk 2 above)
Comment (crazygao) [reply src/promptflow/promptflow/_sdk/entities/_flow.py:201]: OKay, of course.

File: src/promptflow/promptflow/_sdk/entities/_flow.py [hunk 6]
```diff
@@ -167,6 +168,11 @@ def _is_eager_flow(cls, data: dict):
         # If entry specified, it's an eager flow.
         return data.get("entry")
 
+    @classmethod
+    def _is_async_flow(cls, kwargs):
+        """Check if the flow is an async flow. Use field 'entry' to determine."""
+        return kwargs.pop("async_call", False)
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/entities/_flow.py:174]: maybe just async?

File: src/promptflow/promptflow/_sdk/entities/_flow.py [hunk 7]
```diff
@@ -167,6 +168,11 @@ def _is_eager_flow(cls, data: dict):
         # If entry specified, it's an eager flow.
         return data.get("entry")
 
+    @classmethod
+    def _is_async_flow(cls, kwargs):
+        """Check if the flow is an async flow. Use field 'entry' to determine."""
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/entities/_flow.py:173]: why say:  Use field 'entry' to determine

Comment (crazygao) [reply src/promptflow/promptflow/_sdk/entities/_flow.py:173]: Copied method, wrong in comment.

File: src/promptflow/promptflow/_sdk/entities/_flow.py (hunk 6 above)
Comment (crazygao) [reply src/promptflow/promptflow/_sdk/entities/_flow.py:174]: async is keyword in python.

File: src/promptflow/promptflow/_sdk/entities/_flow.py (hunk 5 above)
Comment (crazygao) [reply src/promptflow/promptflow/_sdk/entities/_flow.py:204]: Good Catch.

File: src/promptflow/promptflow/_sdk/entities/_flow.py [hunk 8]
```diff
@@ -102,7 +103,7 @@ def __init__(self, *, data: dict, code: Path, path: Path, **kwargs):
         self._path = Path(path).resolve()
         # hash of flow's entry file, used to skip invoke if entry file is not changed
         self._content_hash = kwargs.pop("content_hash", None)
-        super().__init__(**kwargs)
+        super().__init__()
```
Comment (crazygao) [inline src/promptflow/promptflow/_sdk/entities/_flow.py:106]: super().__init__ doesn't accept kwargs.

**Pull request closed without merging**